# About Dataset
This dataset contains information about used cars.
This data can be used for a lot of purposes such as price prediction to exemplify the use of linear regression in Machine Learning.




In [1]:
#import imp liabrary
import pandas as pd
import numpy as np

we have total 3 datasets .
lets import them

In [2]:
#load the dataset 
data1 = pd.read_csv('car data.csv')
data2 = pd.read_csv('CAR DETAILS FROM CAR DEKHO.csv')
data3 = pd.read_csv('Car details v3.csv')

### data1
for first dataset

In [3]:
data1

Car_Name  Year  Selling_Price  Present_Price  Kms_Driven Fuel_Type  \
0       ritz  2014           3.35           5.59       27000    Petrol   
1        sx4  2013           4.75           9.54       43000    Diesel   
2       ciaz  2017           7.25           9.85        6900    Petrol   
3    wagon r  2011           2.85           4.15        5200    Petrol   
4      swift  2014           4.60           6.87       42450    Diesel   
..       ...   ...            ...            ...         ...       ...   
296     city  2016           9.50          11.60       33988    Diesel   
297     brio  2015           4.00           5.90       60000    Petrol   
298     city  2009           3.35          11.00       87934    Petrol   
299     city  2017          11.50          12.50        9000    Diesel   
300     brio  2016           5.30           5.90        5464    Petrol   

    Seller_Type Transmission  Owner  
0        Dealer       Manual      0  
1        Dealer       Manual      0  
2        Dealer       Manual      0  
3        Dealer       Manual      0  
4        Dealer       Manual      0  
..          ...          ...    ...  
296      Dealer       Manual      0  
297      Dealer       Manual      0  
298      Dealer       Manual      0  
299      Dealer       Manual      0  
300      Dealer       Manual      0  

[301 rows x 9 columns]

In [4]:
data1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 301 entries, 0 to 300
Data columns (total 9 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Car_Name       301 non-null    object 
 1   Year           301 non-null    int64  
 2   Selling_Price  301 non-null    float64
 3   Present_Price  301 non-null    float64
 4   Kms_Driven     301 non-null    int64  
 5   Fuel_Type      301 non-null    object 
 6   Seller_Type    301 non-null    object 
 7   Transmission   301 non-null    object 
 8   Owner          301 non-null    int64  
dtypes: float64(2), int64(3), object(4)
memory usage: 21.3+ KB


In [5]:
data1_colunm = ['Car_Name','Year','Selling_Price','Present_Price','Kms_Driven','Fuel_Type' ,'Seller_Type','Transmission','Owner']

In [6]:
data1_colunm

['Car_Name',
 'Year',
 'Selling_Price',
 'Present_Price',
 'Kms_Driven',
 'Fuel_Type',
 'Seller_Type',
 'Transmission',
 'Owner']

In [7]:
carfromd1 = data1['Car_Name'].value_counts()

# It contains bike also
In data1 there are total 98 types of cars mostly cars are 
1 . city
2 . corolla altis
3 . verna
4 . fortuner

# data2



In [8]:
data2

name  year  selling_price  km_driven  \
0                           Maruti 800 AC  2007          60000      70000   
1                Maruti Wagon R LXI Minor  2007         135000      50000   
2                    Hyundai Verna 1.6 SX  2012         600000     100000   
3                  Datsun RediGO T Option  2017         250000      46000   
4                   Honda Amaze VX i-DTEC  2014         450000     141000   
...                                   ...   ...            ...        ...   
4335  Hyundai i20 Magna 1.4 CRDi (Diesel)  2014         409999      80000   
4336           Hyundai i20 Magna 1.4 CRDi  2014         409999      80000   
4337                  Maruti 800 AC BSIII  2009         110000      83000   
4338     Hyundai Creta 1.6 CRDi SX Option  2016         865000      90000   
4339                     Renault KWID RXT  2016         225000      40000   

        fuel seller_type transmission         owner  
0     Petrol  Individual       Manual   First Owner  
1     Petrol  Individual       Manual   First Owner  
2     Diesel  Individual       Manual   First Owner  
3     Petrol  Individual       Manual   First Owner  
4     Diesel  Individual       Manual  Second Owner  
...      ...         ...          ...           ...  
4335  Diesel  Individual       Manual  Second Owner  
4336  Diesel  Individual       Manual  Second Owner  
4337  Petrol  Individual       Manual  Second Owner  
4338  Diesel  Individual       Manual   First Owner  
4339  Petrol  Individual       Manual   First Owner  

[4340 rows x 8 columns]

In [9]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4340 entries, 0 to 4339
Data columns (total 8 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   name           4340 non-null   object
 1   year           4340 non-null   int64 
 2   selling_price  4340 non-null   int64 
 3   km_driven      4340 non-null   int64 
 4   fuel           4340 non-null   object
 5   seller_type    4340 non-null   object
 6   transmission   4340 non-null   object
 7   owner          4340 non-null   object
dtypes: int64(3), object(5)
memory usage: 271.4+ KB


In [10]:
carlist2 = data2['name'].value_counts()
carlist2 = carlist2.to_dict()
carlist2

{'Maruti Swift Dzire VDI': 69,
 'Maruti Alto 800 LXI': 59,
 'Maruti Alto LXi': 47,
 'Maruti Alto LX': 35,
 'Hyundai EON Era Plus': 35,
 'Maruti Swift VDI BSIV': 29,
 'Maruti Wagon R VXI BS IV': 29,
 'Maruti Swift VDI': 27,
 'Hyundai EON Magna Plus': 24,
 'Maruti Wagon R LXI Minor': 24,
 'Maruti 800 AC': 23,
 'Maruti Wagon R LXI': 23,
 'Maruti Ritz VDi': 22,
 'Mahindra XUV500 W8 2WD': 22,
 'Hyundai i10 Magna': 22,
 'Hyundai Santro Xing GLS': 21,
 'Renault KWID RXT': 21,
 'Maruti Alto K10 VXI': 21,
 'Hyundai Creta 1.6 CRDi SX': 19,
 'Renault Duster 85PS Diesel RxL': 17,
 'Renault KWID 1.0 RXT Optional': 17,
 'Chevrolet Beat Diesel LT': 17,
 'Maruti SX4 Vxi BSIV': 16,
 'Chevrolet Beat Diesel LS': 16,
 'Hyundai Verna 1.6 SX': 15,
 'Chevrolet Spark 1.0 LS': 15,
 'Tata Indica GLS BS IV': 15,
 'Hyundai Verna 1.6 SX CRDi (O)': 14,
 'Maruti Swift VXI': 14,
 'Maruti S-Cross Zeta DDiS 200 SH': 14,
 'Maruti Ertiga VXI': 14,
 'Maruti Alto LXi BSIII': 14,
 'Maruti Alto 800 VXI': 14,
 'Ford Figo Dies

In [11]:
data2['year'].describe()

count    4340.000000
mean     2013.090783
std         4.215344
min      1992.000000
25%      2011.000000
50%      2014.000000
75%      2016.000000
max      2020.000000
Name: year, dtype: float64

In [12]:
data2[data2["year"] <= 2007]


name  year  selling_price  km_driven    fuel  \
0                Maruti 800 AC  2007          60000      70000  Petrol   
1     Maruti Wagon R LXI Minor  2007         135000      50000  Petrol   
5         Maruti Alto LX BSIII  2007         140000     125000  Petrol   
13               Maruti 800 AC  2007          60000      70000  Petrol   
14    Maruti Wagon R LXI Minor  2007         135000      50000  Petrol   
...                        ...   ...            ...        ...     ...   
4302        BMW 7 Series 730Ld  2006        1050000      30000  Diesel   
4318           Maruti Alto STD  2005          95000      50000  Petrol   
4321            Maruti Alto LX  2006          75000     130000  Petrol   
4325           Maruti Alto LXi  2000          65000      90000  Petrol   
4326        Honda City 1.5 GXI  2005          65000     150000  Petrol   

     seller_type transmission         owner  
0     Individual       Manual   First Owner  
1     Individual       Manual   First Owner  
5     Individual       Manual   First Owner  
13    Individual       Manual   First Owner  
14    Individual       Manual   First Owner  
...          ...          ...           ...  
4302      Dealer    Automatic   First Owner  
4318  Individual       Manual  Second Owner  
4321  Individual       Manual   First Owner  
4325  Individual       Manual  Second Owner  
4326  Individual       Manual   Third Owner  

[476 rows x 8 columns]

# data3

In [13]:
data3

name  year  selling_price  km_driven    fuel  \
0           Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1     Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2         Honda City 2017-2020 EXi  2006         158000     140000  Petrol   
3        Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4           Maruti Swift VXI BSIII  2007         130000     120000  Petrol   
...                            ...   ...            ...        ...     ...   
8123             Hyundai i20 Magna  2013         320000     110000  Petrol   
8124         Hyundai Verna CRDi SX  2007         135000     119000  Diesel   
8125        Maruti Swift Dzire ZDi  2009         382000     120000  Diesel   
8126               Tata Indigo CR4  2013         290000      25000  Diesel   
8127               Tata Indigo CR4  2013         290000      25000  Diesel   

     seller_type transmission                 owner     mileage   engine  \
0     Individual       Manual           First Owner   23.4 kmpl  1248 CC   
1     Individual       Manual          Second Owner  21.14 kmpl  1498 CC   
2     Individual       Manual           Third Owner   17.7 kmpl  1497 CC   
3     Individual       Manual           First Owner   23.0 kmpl  1396 CC   
4     Individual       Manual           First Owner   16.1 kmpl  1298 CC   
...          ...          ...                   ...         ...      ...   
8123  Individual       Manual           First Owner   18.5 kmpl  1197 CC   
8124  Individual       Manual  Fourth & Above Owner   16.8 kmpl  1493 CC   
8125  Individual       Manual           First Owner   19.3 kmpl  1248 CC   
8126  Individual       Manual           First Owner  23.57 kmpl  1396 CC   
8127  Individual       Manual           First Owner  23.57 kmpl  1396 CC   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
2         78 bhp      12.7@ 2,700(kgm@ rpm)    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
...          ...                        ...    ...  
8123   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125    73.9 bhp             190Nm@ 2000rpm    5.0  
8126      70 bhp        140Nm@ 1800-3000rpm    5.0  
8127      70 bhp        140Nm@ 1800-3000rpm    5.0  

[8128 rows x 13 columns]

In [14]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8128 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           8128 non-null   object 
 1   year           8128 non-null   int64  
 2   selling_price  8128 non-null   int64  
 3   km_driven      8128 non-null   int64  
 4   fuel           8128 non-null   object 
 5   seller_type    8128 non-null   object 
 6   transmission   8128 non-null   object 
 7   owner          8128 non-null   object 
 8   mileage        7907 non-null   object 
 9   engine         7907 non-null   object 
 10  max_power      7913 non-null   object 
 11  torque         7906 non-null   object 
 12  seats          7907 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 825.6+ KB


# EDA

In [15]:
#delete older car 15 year
data3 = data3[data3["year"] >= 2007]

In [16]:
data3

name  year  selling_price  km_driven    fuel  \
0            Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1      Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
3         Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4            Maruti Swift VXI BSIII  2007         130000     120000  Petrol   
5     Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000  Petrol   
...                             ...   ...            ...        ...     ...   
8123              Hyundai i20 Magna  2013         320000     110000  Petrol   
8124          Hyundai Verna CRDi SX  2007         135000     119000  Diesel   
8125         Maruti Swift Dzire ZDi  2009         382000     120000  Diesel   
8126                Tata Indigo CR4  2013         290000      25000  Diesel   
8127                Tata Indigo CR4  2013         290000      25000  Diesel   

     seller_type transmission                 owner     mileage   engine  \
0     Individual       Manual           First Owner   23.4 kmpl  1248 CC   
1     Individual       Manual          Second Owner  21.14 kmpl  1498 CC   
3     Individual       Manual           First Owner   23.0 kmpl  1396 CC   
4     Individual       Manual           First Owner   16.1 kmpl  1298 CC   
5     Individual       Manual           First Owner  20.14 kmpl  1197 CC   
...          ...          ...                   ...         ...      ...   
8123  Individual       Manual           First Owner   18.5 kmpl  1197 CC   
8124  Individual       Manual  Fourth & Above Owner   16.8 kmpl  1493 CC   
8125  Individual       Manual           First Owner   19.3 kmpl  1248 CC   
8126  Individual       Manual           First Owner  23.57 kmpl  1396 CC   
8127  Individual       Manual           First Owner  23.57 kmpl  1396 CC   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
5      81.86 bhp          113.75nm@ 4000rpm    5.0  
...          ...                        ...    ...  
8123   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125    73.9 bhp             190Nm@ 2000rpm    5.0  
8126      70 bhp        140Nm@ 1800-3000rpm    5.0  
8127      70 bhp        140Nm@ 1800-3000rpm    5.0  

[7688 rows x 13 columns]

In [17]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7688 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           7688 non-null   object 
 1   year           7688 non-null   int64  
 2   selling_price  7688 non-null   int64  
 3   km_driven      7688 non-null   int64  
 4   fuel           7688 non-null   object 
 5   seller_type    7688 non-null   object 
 6   transmission   7688 non-null   object 
 7   owner          7688 non-null   object 
 8   mileage        7563 non-null   object 
 9   engine         7563 non-null   object 
 10  max_power      7569 non-null   object 
 11  torque         7563 non-null   object 
 12  seats          7563 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 840.9+ KB


In [18]:
data3.describe()

year  selling_price     km_driven        seats
count  7688.000000   7.688000e+03  7.688000e+03  7563.000000
mean   2014.392300   6.676533e+05  6.817190e+04     5.428401
std       3.227404   8.190050e+05  5.644926e+04     0.950506
min    2007.000000   3.500000e+04  1.000000e+00     2.000000
25%    2012.000000   2.900000e+05  3.200000e+04     5.000000
50%    2015.000000   4.700000e+05  6.000000e+04     5.000000
75%    2017.000000   7.000000e+05  9.100000e+04     5.000000
max    2020.000000   1.000000e+07  2.360457e+06    14.000000

In [19]:
carlist3 = data3['name'].value_counts()
carlist3 = carlist3.to_dict()
carlist3
print(f'We have information of {len(carlist3)} total vehicals')

We have information of 1913 total vehicals


In [20]:
data3['fuel'].value_counts()


Diesel    4292
Petrol    3306
CNG         55
LPG         35
Name: fuel, dtype: int64

In [21]:
null_values = data3[data3["mileage"].isnull()]
null_values

name  year  selling_price  km_driven    fuel  \
13           Maruti Swift 1.3 VXi  2007         200000      80000  Petrol   
87    Maruti Swift VDI BSIV W ABS  2015         475000      78000  Diesel   
119         Maruti Swift VDI BSIV  2010         300000     120000  Diesel   
138       BMW X1 sDrive 20D xLine  2017        2150000      27000  Diesel   
200               Toyota Etios GD  2012         235000     120000  Diesel   
...                           ...   ...            ...        ...     ...   
7454    Ford Figo Aspire Facelift  2018         624000      14681  Diesel   
7528            Maruti Estilo LXI  2008         160000      56000  Petrol   
7651        Maruti Swift VDI BSIV  2014         217000      50000  Diesel   
7730      Maruti Omni E 8 Str STD  2008         125000      70000  Petrol   
8068    Ford Figo Aspire Facelift  2017         580000     165000  Diesel   

     seller_type transmission                 owner mileage engine max_power  \
13    Individual       Manual          Second Owner     NaN    NaN       NaN   
87        Dealer       Manual           First Owner     NaN    NaN       NaN   
119   Individual       Manual          Second Owner     NaN    NaN       NaN   
138       Dealer    Automatic           First Owner     NaN    NaN       NaN   
200   Individual       Manual           First Owner     NaN    NaN       NaN   
...          ...          ...                   ...     ...    ...       ...   
7454      Dealer       Manual           First Owner     NaN    NaN       NaN   
7528  Individual       Manual           Third Owner     NaN    NaN       NaN   
7651  Individual       Manual           First Owner     NaN    NaN       NaN   
7730  Individual       Manual  Fourth & Above Owner     NaN    NaN       NaN   
8068  Individual       Manual           First Owner     NaN    NaN       NaN   

     torque  seats  
13      NaN    NaN  
87      NaN    NaN  
119     NaN    NaN  
138     NaN    NaN  
200     NaN    NaN  
...     ...    ...  
7454    NaN    NaN  
7528    NaN    NaN  
7651    NaN    NaN  
7730    NaN    NaN  
8068    NaN    NaN  

[125 rows x 13 columns]

In [22]:
null_values.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 125 entries, 13 to 8068
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           125 non-null    object 
 1   year           125 non-null    int64  
 2   selling_price  125 non-null    int64  
 3   km_driven      125 non-null    int64  
 4   fuel           125 non-null    object 
 5   seller_type    125 non-null    object 
 6   transmission   125 non-null    object 
 7   owner          125 non-null    object 
 8   mileage        0 non-null      object 
 9   engine         0 non-null      object 
 10  max_power      6 non-null      object 
 11  torque         0 non-null      object 
 12  seats          0 non-null      float64
dtypes: float64(1), int64(3), object(9)
memory usage: 13.7+ KB


######  125 values are missed in to same rows ,collumns are mileage	,engine	,torque	,seats and 	max_power 119

In [23]:
#droping NaN values
new_data3 = data3.dropna()
new_data3

name  year  selling_price  km_driven    fuel  \
0            Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1      Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
3         Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   
4            Maruti Swift VXI BSIII  2007         130000     120000  Petrol   
5     Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000  Petrol   
...                             ...   ...            ...        ...     ...   
8123              Hyundai i20 Magna  2013         320000     110000  Petrol   
8124          Hyundai Verna CRDi SX  2007         135000     119000  Diesel   
8125         Maruti Swift Dzire ZDi  2009         382000     120000  Diesel   
8126                Tata Indigo CR4  2013         290000      25000  Diesel   
8127                Tata Indigo CR4  2013         290000      25000  Diesel   

     seller_type transmission                 owner     mileage   engine  \
0     Individual       Manual           First Owner   23.4 kmpl  1248 CC   
1     Individual       Manual          Second Owner  21.14 kmpl  1498 CC   
3     Individual       Manual           First Owner   23.0 kmpl  1396 CC   
4     Individual       Manual           First Owner   16.1 kmpl  1298 CC   
5     Individual       Manual           First Owner  20.14 kmpl  1197 CC   
...          ...          ...                   ...         ...      ...   
8123  Individual       Manual           First Owner   18.5 kmpl  1197 CC   
8124  Individual       Manual  Fourth & Above Owner   16.8 kmpl  1493 CC   
8125  Individual       Manual           First Owner   19.3 kmpl  1248 CC   
8126  Individual       Manual           First Owner  23.57 kmpl  1396 CC   
8127  Individual       Manual           First Owner  23.57 kmpl  1396 CC   

       max_power                     torque  seats  
0         74 bhp             190Nm@ 2000rpm    5.0  
1     103.52 bhp        250Nm@ 1500-2500rpm    5.0  
3         90 bhp   22.4 kgm at 1750-2750rpm    5.0  
4       88.2 bhp      11.5@ 4,500(kgm@ rpm)    5.0  
5      81.86 bhp          113.75nm@ 4000rpm    5.0  
...          ...                        ...    ...  
8123   82.85 bhp           113.7Nm@ 4000rpm    5.0  
8124     110 bhp  24@ 1,900-2,750(kgm@ rpm)    5.0  
8125    73.9 bhp             190Nm@ 2000rpm    5.0  
8126      70 bhp        140Nm@ 1800-3000rpm    5.0  
8127      70 bhp        140Nm@ 1800-3000rpm    5.0  

[7563 rows x 13 columns]

In [24]:
new_data3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7563 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           7563 non-null   object 
 1   year           7563 non-null   int64  
 2   selling_price  7563 non-null   int64  
 3   km_driven      7563 non-null   int64  
 4   fuel           7563 non-null   object 
 5   seller_type    7563 non-null   object 
 6   transmission   7563 non-null   object 
 7   owner          7563 non-null   object 
 8   mileage        7563 non-null   object 
 9   engine         7563 non-null   object 
 10  max_power      7563 non-null   object 
 11  torque         7563 non-null   object 
 12  seats          7563 non-null   float64
dtypes: float64(1), int64(3), object(9)
memory usage: 827.2+ KB


In [25]:
new_data3.describe()

year  selling_price     km_driven        seats
count  7563.000000   7.563000e+03  7.563000e+03  7563.000000
mean   2014.450615   6.735058e+05  6.778021e+04     5.428401
std       3.204613   8.237217e+05  5.660461e+04     0.950506
min    2007.000000   3.500000e+04  1.000000e+00     2.000000
25%    2012.000000   2.900000e+05  3.180000e+04     5.000000
50%    2015.000000   4.750000e+05  6.000000e+04     5.000000
75%    2017.000000   7.000000e+05  9.000000e+04     5.000000
max    2020.000000   1.000000e+07  2.360457e+06    14.000000

In [26]:
new_data3['seats'] = new_data3['seats'].astype(int)

C:\Users\UUSHA INFO\AppData\Local\Temp\ipykernel_6876\1257093897.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data3['seats'] = new_data3['seats'].astype(int)


In [27]:
new_data3.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7563 entries, 0 to 8127
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   name           7563 non-null   object
 1   year           7563 non-null   int64 
 2   selling_price  7563 non-null   int64 
 3   km_driven      7563 non-null   int64 
 4   fuel           7563 non-null   object
 5   seller_type    7563 non-null   object
 6   transmission   7563 non-null   object
 7   owner          7563 non-null   object
 8   mileage        7563 non-null   object
 9   engine         7563 non-null   object
 10  max_power      7563 non-null   object
 11  torque         7563 non-null   object
 12  seats          7563 non-null   int32 
dtypes: int32(1), int64(3), object(9)
memory usage: 797.7+ KB


In [28]:
ndata3 =new_data3.drop_duplicates()

In [29]:
ndata3

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         450000     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
3             Hyundai i20 Sportz Diesel  2010         225000     127000   
4                Maruti Swift VXI BSIII  2007         130000     120000   
5         Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000   
...                                 ...   ...            ...        ...   
8121  Maruti Wagon R VXI BS IV with ABS  2013         260000      50000   
8122         Hyundai i20 Magna 1.4 CRDi  2014         475000      80000   
8123                  Hyundai i20 Magna  2013         320000     110000   
8124              Hyundai Verna CRDi SX  2007         135000     119000   
8125             Maruti Swift Dzire ZDi  2009         382000     120000   

        fuel seller_type transmission                 owner     mileage  \
0     Diesel  Individual       Manual           First Owner   23.4 kmpl   
1     Diesel  Individual       Manual          Second Owner  21.14 kmpl   
3     Diesel  Individual       Manual           First Owner   23.0 kmpl   
4     Petrol  Individual       Manual           First Owner   16.1 kmpl   
5     Petrol  Individual       Manual           First Owner  20.14 kmpl   
...      ...         ...          ...                   ...         ...   
8121  Petrol  Individual       Manual          Second Owner   18.9 kmpl   
8122  Diesel  Individual       Manual          Second Owner  22.54 kmpl   
8123  Petrol  Individual       Manual           First Owner   18.5 kmpl   
8124  Diesel  Individual       Manual  Fourth & Above Owner   16.8 kmpl   
8125  Diesel  Individual       Manual           First Owner   19.3 kmpl   

       engine   max_power                     torque  seats  
0     1248 CC      74 bhp             190Nm@ 2000rpm      5  
1     1498 CC  103.52 bhp        250Nm@ 1500-2500rpm      5  
3     1396 CC      90 bhp   22.4 kgm at 1750-2750rpm      5  
4     1298 CC    88.2 bhp      11.5@ 4,500(kgm@ rpm)      5  
5     1197 CC   81.86 bhp          113.75nm@ 4000rpm      5  
...       ...         ...                        ...    ...  
8121   998 CC    67.1 bhp              90Nm@ 3500rpm      5  
8122  1396 CC   88.73 bhp      219.7Nm@ 1500-2750rpm      5  
8123  1197 CC   82.85 bhp           113.7Nm@ 4000rpm      5  
8124  1493 CC     110 bhp  24@ 1,900-2,750(kgm@ rpm)      5  
8125  1248 CC    73.9 bhp             190Nm@ 2000rpm      5  

[6385 rows x 13 columns]

In [30]:
cars_under_1000_km = ndata3[ndata3['km_driven'] <= 1000]

In [31]:
cars_under_1000_km

name  year  selling_price  km_driven  \
399                  Honda City i-VTEC CVT V  2018        1085000       1000   
2963                    Maruti Swift AMT VXI  2020         654000       1000   
6638          Toyota Innova Crysta 2.4 ZX AT  2020        2300000       1000   
7429                        Maruti Swift LXI  2020         445000       1000   
7492                    Hyundai Santro Magna  2020         500000       1000   
7913  Maruti Eeco 5 STR With AC Plus HTR CNG  2011         209000          1   

        fuel seller_type transmission                 owner     mileage  \
399   Petrol  Individual    Automatic           First Owner   18.0 kmpl   
2963  Petrol  Individual    Automatic           First Owner  21.21 kmpl   
6638  Diesel  Individual    Automatic           First Owner  13.68 kmpl   
7429  Petrol  Individual       Manual           First Owner  21.21 kmpl   
7492  Petrol  Individual       Manual           First Owner   20.3 kmpl   
7913     CNG  Individual       Manual  Fourth & Above Owner  15.1 km/kg   

       engine   max_power               torque  seats  
399   1497 CC   117.6 bhp       145Nm@ 4600rpm      5  
2963  1197 CC   81.80 bhp       113Nm@ 4200rpm      5  
6638  2393 CC  147.51 bhp  360Nm@ 1400-2600rpm      7  
7429  1197 CC   81.80 bhp       113Nm@ 4200rpm      5  
7492  1086 CC   68.07 bhp     99.07Nm@ 4500rpm      5  
7913  1196 CC      73 bhp       101Nm@ 3000rpm      5

some colunms are not numeric format even they represent numerical charectoristics


In [32]:
ndata3.reset_index(inplace=True)

In [33]:
ndata3

index                               name  year  selling_price  \
0         0             Maruti Swift Dzire VDI  2014         450000   
1         1       Skoda Rapid 1.5 TDI Ambition  2014         370000   
2         3          Hyundai i20 Sportz Diesel  2010         225000   
3         4             Maruti Swift VXI BSIII  2007         130000   
4         5      Hyundai Xcent 1.2 VTVT E Plus  2017         440000   
...     ...                                ...   ...            ...   
6380   8121  Maruti Wagon R VXI BS IV with ABS  2013         260000   
6381   8122         Hyundai i20 Magna 1.4 CRDi  2014         475000   
6382   8123                  Hyundai i20 Magna  2013         320000   
6383   8124              Hyundai Verna CRDi SX  2007         135000   
6384   8125             Maruti Swift Dzire ZDi  2009         382000   

      km_driven    fuel seller_type transmission                 owner  \
0        145500  Diesel  Individual       Manual           First Owner   
1        120000  Diesel  Individual       Manual          Second Owner   
2        127000  Diesel  Individual       Manual           First Owner   
3        120000  Petrol  Individual       Manual           First Owner   
4         45000  Petrol  Individual       Manual           First Owner   
...         ...     ...         ...          ...                   ...   
6380      50000  Petrol  Individual       Manual          Second Owner   
6381      80000  Diesel  Individual       Manual          Second Owner   
6382     110000  Petrol  Individual       Manual           First Owner   
6383     119000  Diesel  Individual       Manual  Fourth & Above Owner   
6384     120000  Diesel  Individual       Manual           First Owner   

         mileage   engine   max_power                     torque  seats  
0      23.4 kmpl  1248 CC      74 bhp             190Nm@ 2000rpm      5  
1     21.14 kmpl  1498 CC  103.52 bhp        250Nm@ 1500-2500rpm      5  
2      23.0 kmpl  1396 CC      90 bhp   22.4 kgm at 1750-2750rpm      5  
3      16.1 kmpl  1298 CC    88.2 bhp      11.5@ 4,500(kgm@ rpm)      5  
4     20.14 kmpl  1197 CC   81.86 bhp          113.75nm@ 4000rpm      5  
...          ...      ...         ...                        ...    ...  
6380   18.9 kmpl   998 CC    67.1 bhp              90Nm@ 3500rpm      5  
6381  22.54 kmpl  1396 CC   88.73 bhp      219.7Nm@ 1500-2750rpm      5  
6382   18.5 kmpl  1197 CC   82.85 bhp           113.7Nm@ 4000rpm      5  
6383   16.8 kmpl  1493 CC     110 bhp  24@ 1,900-2,750(kgm@ rpm)      5  
6384   19.3 kmpl  1248 CC    73.9 bhp             190Nm@ 2000rpm      5  

[6385 rows x 14 columns]

In [34]:
try:
    ndata3 = ndata3.drop(columns='index')
    print('deleted!')
except Exception as e:
    print(e)
    

deleted!


In [35]:
ndata3.head(3)

name  year  selling_price  km_driven    fuel  \
0        Maruti Swift Dzire VDI  2014         450000     145500  Diesel   
1  Skoda Rapid 1.5 TDI Ambition  2014         370000     120000  Diesel   
2     Hyundai i20 Sportz Diesel  2010         225000     127000  Diesel   

  seller_type transmission         owner     mileage   engine   max_power  \
0  Individual       Manual   First Owner   23.4 kmpl  1248 CC      74 bhp   
1  Individual       Manual  Second Owner  21.14 kmpl  1498 CC  103.52 bhp   
2  Individual       Manual   First Owner   23.0 kmpl  1396 CC      90 bhp   

                     torque  seats  
0            190Nm@ 2000rpm      5  
1       250Nm@ 1500-2500rpm      5  
2  22.4 kgm at 1750-2750rpm      5

In [36]:
#Slicing unnesesory part
data = ndata3
data['mileage']=ndata3['mileage'].str.replace("kmpl","")
data['engine']=ndata3['engine'].str.replace('CC',' ')
data['max_power']=ndata3['max_power'].str.replace('bhp',' ')
data

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         450000     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2             Hyundai i20 Sportz Diesel  2010         225000     127000   
3                Maruti Swift VXI BSIII  2007         130000     120000   
4         Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000   
...                                 ...   ...            ...        ...   
6380  Maruti Wagon R VXI BS IV with ABS  2013         260000      50000   
6381         Hyundai i20 Magna 1.4 CRDi  2014         475000      80000   
6382                  Hyundai i20 Magna  2013         320000     110000   
6383              Hyundai Verna CRDi SX  2007         135000     119000   
6384             Maruti Swift Dzire ZDi  2009         382000     120000   

        fuel seller_type transmission                 owner mileage  engine  \
0     Diesel  Individual       Manual           First Owner   23.4   1248     
1     Diesel  Individual       Manual          Second Owner  21.14   1498     
2     Diesel  Individual       Manual           First Owner   23.0   1396     
3     Petrol  Individual       Manual           First Owner   16.1   1298     
4     Petrol  Individual       Manual           First Owner  20.14   1197     
...      ...         ...          ...                   ...     ...     ...   
6380  Petrol  Individual       Manual          Second Owner   18.9    998     
6381  Diesel  Individual       Manual          Second Owner  22.54   1396     
6382  Petrol  Individual       Manual           First Owner   18.5   1197     
6383  Diesel  Individual       Manual  Fourth & Above Owner   16.8   1493     
6384  Diesel  Individual       Manual           First Owner   19.3   1248     

     max_power                     torque  seats  
0         74               190Nm@ 2000rpm      5  
1     103.52          250Nm@ 1500-2500rpm      5  
2         90     22.4 kgm at 1750-2750rpm      5  
3       88.2        11.5@ 4,500(kgm@ rpm)      5  
4      81.86            113.75nm@ 4000rpm      5  
...        ...                        ...    ...  
6380    67.1                90Nm@ 3500rpm      5  
6381   88.73        219.7Nm@ 1500-2750rpm      5  
6382   82.85             113.7Nm@ 4000rpm      5  
6383     110    24@ 1,900-2,750(kgm@ rpm)      5  
6384    73.9               190Nm@ 2000rpm      5  

[6385 rows x 13 columns]

##### lets find out what things are have in torque column

In [37]:
torque = data['torque']

In [38]:
Horsepower = 190 * 2000 
Horsepower = Horsepower/7127
Horsepower

53.31836677423881

###### we need to convert it into hp for that
### power = nm*rpm  ,
## hp = power/7127          
## 1 kg-m is equal to 9.80665 newton meter.
#

In [39]:
#........................................................
torque1 = data['torque']
torque1.value_counts()

190Nm@ 2000rpm           472
200Nm@ 1750rpm           398
90Nm@ 3500rpm            321
114Nm@ 4000rpm           156
113Nm@ 4200rpm           139
                        ... 
240Nm@ 1850rpm             1
347Nm@ 4300rpm             1
130@ 2500(kgm@ rpm)        1
219.9Nm@ 1750-2750rpm      1
96  Nm at 3000  rpm        1
Name: torque, Length: 415, dtype: int64

In [40]:
torque1 = torque1.str.replace("Nm at",'@')
torque1.value_counts()

190Nm@ 2000rpm           472
200Nm@ 1750rpm           398
90Nm@ 3500rpm            321
114Nm@ 4000rpm           156
113Nm@ 4200rpm           139
                        ... 
240Nm@ 1850rpm             1
347Nm@ 4300rpm             1
130@ 2500(kgm@ rpm)        1
219.9Nm@ 1750-2750rpm      1
96  @ 3000  rpm            1
Name: torque, Length: 415, dtype: int64

In [41]:
torque1 = torque1.str.replace("Nm@",'')
torque1 = torque1.str.replace("nm@",'')
torque1 = torque1.str.replace("nm",'')
torque1 = torque1.str.replace("Nm",'')
torque1 = torque1.str.replace("NM",'')
torque1 = torque1.str.replace("NM@",'')
torque1 = torque1.str.replace("NM@",'')
torque1 = torque1.str.replace("kgm@",'@')
torque1 = torque1.str.replace(" kgm at",'@')
torque1 = torque1.str.replace(" KGM at",'@')
torque1 = torque1.str.replace(" @",'@')
torque1 = torque1.str.replace("KGM@",'@')


torque1.value_counts()

190 2000rpm              472
200 1750rpm              398
90 3500rpm               322
114 4000rpm              156
113 4200rpm              139
                        ... 
160 1750rpm                1
115 3200rpm                1
215 3600rpm                1
380(38.7kgm)@ 2500rpm      1
96 @ 3000  rpm             1
Name: torque, Length: 405, dtype: int64

In [42]:
torque1 = torque1.str.split(n=1,expand=True)

In [43]:
torque1

0                   1
0        190             2000rpm
1        250        1500-2500rpm
2      22.4@        1750-2750rpm
3      11.5@        4,500(@ rpm)
4     113.75             4000rpm
...      ...                 ...
6380      90             3500rpm
6381   219.7        1500-2750rpm
6382   113.7             4000rpm
6383     24@  1,900-2,750(@ rpm)
6384     190             2000rpm

[6385 rows x 2 columns]

In [44]:
torque1.columns =['Nm','rpm']
torque1

Nm                 rpm
0        190             2000rpm
1        250        1500-2500rpm
2      22.4@        1750-2750rpm
3      11.5@        4,500(@ rpm)
4     113.75             4000rpm
...      ...                 ...
6380      90             3500rpm
6381   219.7        1500-2750rpm
6382   113.7             4000rpm
6383     24@  1,900-2,750(@ rpm)
6384     190             2000rpm

[6385 rows x 2 columns]

In [45]:

l =torque1.Nm.value_counts()
print(l)

200      594
190      541
90       323
114      227
113      175
        ... 
22.9@      1
23.6@      1
35.7@      1
6@         1
436.4      1
Name: Nm, Length: 253, dtype: int64


In [46]:
from fuzzywuzzy import process
query ='380(38.7kgm)@'
Nm= torque1['Nm']
re=process.extractOne(query,Nm)
re

C:\Users\UUSHA INFO\anaconda3\lib\site-packages\fuzzywuzzy\fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


('380(38.7kgm)@', 100, 677)

In [47]:
re=process.extractOne('110(11.2)@',Nm)
re

('110(11.2)@', 100, 1673)

In [48]:
Nm[677]= '38.7@'
Nm[1673]='11.2@'

In [49]:
Nm

0          190
1          250
2        22.4@
3        11.5@
4       113.75
         ...  
6380        90
6381     219.7
6382     113.7
6383       24@
6384       190
Name: Nm, Length: 6385, dtype: object

In [50]:
Nm.value_counts()

200      594
190      541
90       323
114      227
113      175
        ... 
185        1
380@       1
22.9@      1
23.6@      1
436.4      1
Name: Nm, Length: 252, dtype: int64

In [51]:
Nm =Nm.str.replace("@"," 9.80665")


In [52]:
Nm =Nm.str.split(' ',expand=True)

In [53]:
Nm

0        1
0        190     None
1        250     None
2       22.4  9.80665
3       11.5  9.80665
4     113.75     None
...      ...      ...
6380      90     None
6381   219.7     None
6382   113.7     None
6383      24  9.80665
6384     190     None

[6385 rows x 2 columns]

In [54]:
Nm.columns=['first_col','last_col']
Nm

first_col last_col
0          190     None
1          250     None
2         22.4  9.80665
3         11.5  9.80665
4       113.75     None
...        ...      ...
6380        90     None
6381     219.7     None
6382     113.7     None
6383        24  9.80665
6384       190     None

[6385 rows x 2 columns]

In [55]:
Nm['last_col']=Nm['last_col'].fillna(1)
Nm

first_col last_col
0          190        1
1          250        1
2         22.4  9.80665
3         11.5  9.80665
4       113.75        1
...        ...      ...
6380        90        1
6381     219.7        1
6382     113.7        1
6383        24  9.80665
6384       190        1

[6385 rows x 2 columns]

In [56]:
Nm["first_col"] =Nm["first_col"].astype(float)
Nm["last_col"] =Nm["last_col"].astype(float)
Nm.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6385 entries, 0 to 6384
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   first_col  6385 non-null   float64
 1   last_col   6385 non-null   float64
dtypes: float64(2)
memory usage: 99.9 KB


In [57]:
Nm = (Nm.first_col*Nm.last_col).astype(int)

In [58]:
Nm

0       190
1       250
2       219
3       112
4       113
       ... 
6380     90
6381    219
6382    113
6383    235
6384    190
Length: 6385, dtype: int32

In [59]:
data['Nm'] = Nm

In [60]:
data

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         450000     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2             Hyundai i20 Sportz Diesel  2010         225000     127000   
3                Maruti Swift VXI BSIII  2007         130000     120000   
4         Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000   
...                                 ...   ...            ...        ...   
6380  Maruti Wagon R VXI BS IV with ABS  2013         260000      50000   
6381         Hyundai i20 Magna 1.4 CRDi  2014         475000      80000   
6382                  Hyundai i20 Magna  2013         320000     110000   
6383              Hyundai Verna CRDi SX  2007         135000     119000   
6384             Maruti Swift Dzire ZDi  2009         382000     120000   

        fuel seller_type transmission                 owner mileage  engine  \
0     Diesel  Individual       Manual           First Owner   23.4   1248     
1     Diesel  Individual       Manual          Second Owner  21.14   1498     
2     Diesel  Individual       Manual           First Owner   23.0   1396     
3     Petrol  Individual       Manual           First Owner   16.1   1298     
4     Petrol  Individual       Manual           First Owner  20.14   1197     
...      ...         ...          ...                   ...     ...     ...   
6380  Petrol  Individual       Manual          Second Owner   18.9    998     
6381  Diesel  Individual       Manual          Second Owner  22.54   1396     
6382  Petrol  Individual       Manual           First Owner   18.5   1197     
6383  Diesel  Individual       Manual  Fourth & Above Owner   16.8   1493     
6384  Diesel  Individual       Manual           First Owner   19.3   1248     

     max_power                     torque  seats   Nm  
0         74               190Nm@ 2000rpm      5  190  
1     103.52          250Nm@ 1500-2500rpm      5  250  
2         90     22.4 kgm at 1750-2750rpm      5  219  
3       88.2        11.5@ 4,500(kgm@ rpm)      5  112  
4      81.86            113.75nm@ 4000rpm      5  113  
...        ...                        ...    ...  ...  
6380    67.1                90Nm@ 3500rpm      5   90  
6381   88.73        219.7Nm@ 1500-2750rpm      5  219  
6382   82.85             113.7Nm@ 4000rpm      5  113  
6383     110    24@ 1,900-2,750(kgm@ rpm)      5  235  
6384    73.9               190Nm@ 2000rpm      5  190  

[6385 rows x 14 columns]

In [61]:
#now for RPM
RPM = torque1['rpm']
RPM

0                  2000rpm
1             1500-2500rpm
2             1750-2750rpm
3             4,500(@ rpm)
4                  4000rpm
               ...        
6380               3500rpm
6381          1500-2750rpm
6382               4000rpm
6383    1,900-2,750(@ rpm)
6384               2000rpm
Name: rpm, Length: 6385, dtype: object

In [62]:
r_list = RPM.value_counts()
r_list

4000rpm               793
2000rpm               707
3500rpm               623
1750rpm               503
3000rpm               243
                     ... 
1740rpm                 1
4,000(@ rpm)            1
1,950-4,700(@ rpm)      1
4800                    1
@ 3000  rpm             1
Name: rpm, Length: 170, dtype: int64

In [63]:
RPM = RPM.str.replace("rpm",'')
RPM = RPM.str.replace("[(@ rpm)]",'')
RPM = RPM.str.replace("@",'')
RPM = RPM.str.replace("kgat",'')
RPM

C:\Users\UUSHA INFO\AppData\Local\Temp\ipykernel_6876\2451279388.py:2: FutureWarning: The default value of regex will change from True to False in a future version.
  RPM = RPM.str.replace("[(@ rpm)]",'')


0              2000
1         1500-2500
2         1750-2750
3             4,500
4              4000
           ...     
6380           3500
6381      1500-2750
6382           4000
6383    1,900-2,750
6384           2000
Name: rpm, Length: 6385, dtype: object

In [64]:
pd.options.display.max_rows=200
RPM.value_counts()

4000            793
2000            707
3500            623
1750            509
3000            247
4200            206
1600-2800       180
1500-2750       171
2500            164
4500            151
1750-3000       139
1900-2750       132
1400-3400       130
1750-2500       130
1800-2800       128
1400-2200       115
1500-2500        96
4800             80
4250             70
4386             70
1440-1500        68
1750-2750        68
4400             63
1700-2700        63
1800-3000        61
1750-3250        60
1500-3000        55
1750-2250        54
4600             52
2,500            41
1800-2400        40
1900             38
2250             36
1400-2800        33
2,000            33
4,000            29
1400-2600        27
1200-3600        26
1500-4000        24
1800-2200        23
1,900-2,750      23
4850             23
3500-3600        22
5000             21
3750             20
1800-2000        20
1800             20
3,000            20
2,700            17
1750-4000        17


KGMat
RPM
~
+/

In [65]:
RPM = RPM.str.replace("KGMat",'')
RPM = RPM.str.replace("RPM",'')
RPM = RPM.str.replace("~",'')
RPM = RPM.str.replace(",",'')
RPM = RPM.str.replace("[+/]",'')
RPM = RPM.str.replace("[-]",' ')
RPM = RPM.str.replace("at",'')

pd.options.display.max_rows=200
RPM.value_counts()

C:\Users\UUSHA INFO\AppData\Local\Temp\ipykernel_6876\2333069497.py:5: FutureWarning: The default value of regex will change from True to False in a future version.
  RPM = RPM.str.replace("[+/]",'')
C:\Users\UUSHA INFO\AppData\Local\Temp\ipykernel_6876\2333069497.py:6: FutureWarning: The default value of regex will change from True to False in a future version.
  RPM = RPM.str.replace("[-]",' ')


4000         822
2000         741
3500         635
1750         509
3000         270
4200         208
2500         205
1600 2800    180
1500 2750    171
4500         166
1900 2750    165
1750 3000    141
1750 2500    132
1400 3400    130
1800 2800    128
1400 2200    115
1500 2500     96
4800          80
4250          71
4386          70
1750 2750     68
1440 1500     68
4400          66
1800 3000     64
1700 2700     63
1750 3250     60
1500 3000     57
1750 2250     54
4600          52
1900          49
1800 2400     40
2250          36
1400 2800     33
1800          27
1400 2600     27
1200 3600     26
2700          26
1500 4000     24
4850          23
1800 2200     23
3750          22
3500 3600     22
5000          21
1800 2000     20
2600          20
4100          20
1750 4000     17
4300          17
2000 2750     16
3000 4300     15
3800          15
1200 3400     14
1600 2400     14
1750 2800     13
1700 2200     13
4000 500      12
1600 3000     11
1000 2000     10
3400          

In [66]:
RPM = RPM.str.split(' ',expand=True)
RPM

0     1
0     2000  None
1     1500  2500
2     1750  2750
3     4500  None
4     4000  None
...    ...   ...
6380  3500  None
6381  1500  2750
6382  4000  None
6383  1900  2750
6384  2000  None

[6385 rows x 2 columns]

In [67]:
RPM.columns = ['f','l']

In [68]:
RPM['first_col']=RPM['f']
RPM['last_col']=RPM['l'].fillna(RPM.pop('f'))
RPM

l first_col last_col
0     None      2000     2000
1     2500      1500     2500
2     2750      1750     2750
3     None      4500     4500
4     None      4000     4000
...    ...       ...      ...
6380  None      3500     3500
6381  2750      1500     2750
6382  None      4000     4000
6383  2750      1900     2750
6384  None      2000     2000

[6385 rows x 3 columns]

In [69]:
RPM.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6385 entries, 0 to 6384
Data columns (total 3 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   l          2160 non-null   object
 1   first_col  6381 non-null   object
 2   last_col   6381 non-null   object
dtypes: object(3)
memory usage: 149.8+ KB


In [70]:
RPM['first_col']=RPM['first_col'].astype(float)
RPM['last_col']=RPM['last_col'].astype(float)

In [71]:
RPM['rpm'] = ((RPM.first_col + RPM.last_col)/2)
#RPM['rpm'] = RPM['rpm'].astype("Int64")
 #np.floor(pd.to_numeric(df['A'], errors='coerce')).astype('Int64')
RPM['rpm'] = np.floor(pd.to_numeric(RPM['rpm'])).astype('Int64')

In [72]:
RPM['rpm']

0       2000
1       2000
2       2250
3       4500
4       4000
        ... 
6380    3500
6381    2125
6382    4000
6383    2325
6384    2000
Name: rpm, Length: 6385, dtype: Int64

In [73]:
data['rpm'] = RPM['rpm']

In [74]:
data

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         450000     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2             Hyundai i20 Sportz Diesel  2010         225000     127000   
3                Maruti Swift VXI BSIII  2007         130000     120000   
4         Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000   
...                                 ...   ...            ...        ...   
6380  Maruti Wagon R VXI BS IV with ABS  2013         260000      50000   
6381         Hyundai i20 Magna 1.4 CRDi  2014         475000      80000   
6382                  Hyundai i20 Magna  2013         320000     110000   
6383              Hyundai Verna CRDi SX  2007         135000     119000   
6384             Maruti Swift Dzire ZDi  2009         382000     120000   

        fuel seller_type transmission                 owner mileage  engine  \
0     Diesel  Individual       Manual           First Owner   23.4   1248     
1     Diesel  Individual       Manual          Second Owner  21.14   1498     
2     Diesel  Individual       Manual           First Owner   23.0   1396     
3     Petrol  Individual       Manual           First Owner   16.1   1298     
4     Petrol  Individual       Manual           First Owner  20.14   1197     
...      ...         ...          ...                   ...     ...     ...   
6380  Petrol  Individual       Manual          Second Owner   18.9    998     
6381  Diesel  Individual       Manual          Second Owner  22.54   1396     
6382  Petrol  Individual       Manual           First Owner   18.5   1197     
6383  Diesel  Individual       Manual  Fourth & Above Owner   16.8   1493     
6384  Diesel  Individual       Manual           First Owner   19.3   1248     

     max_power                     torque  seats   Nm   rpm  
0         74               190Nm@ 2000rpm      5  190  2000  
1     103.52          250Nm@ 1500-2500rpm      5  250  2000  
2         90     22.4 kgm at 1750-2750rpm      5  219  2250  
3       88.2        11.5@ 4,500(kgm@ rpm)      5  112  4500  
4      81.86            113.75nm@ 4000rpm      5  113  4000  
...        ...                        ...    ...  ...   ...  
6380    67.1                90Nm@ 3500rpm      5   90  3500  
6381   88.73        219.7Nm@ 1500-2750rpm      5  219  2125  
6382   82.85             113.7Nm@ 4000rpm      5  113  4000  
6383     110    24@ 1,900-2,750(kgm@ rpm)      5  235  2325  
6384    73.9               190Nm@ 2000rpm      5  190  2000  

[6385 rows x 15 columns]

In [75]:
data.to_csv('car_price_prediction.csv')

In [76]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6385 entries, 0 to 6384
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   name           6385 non-null   object
 1   year           6385 non-null   int64 
 2   selling_price  6385 non-null   int64 
 3   km_driven      6385 non-null   int64 
 4   fuel           6385 non-null   object
 5   seller_type    6385 non-null   object
 6   transmission   6385 non-null   object
 7   owner          6385 non-null   object
 8   mileage        6385 non-null   object
 9   engine         6385 non-null   object
 10  max_power      6385 non-null   object
 11  torque         6385 non-null   object
 12  seats          6385 non-null   int32 
 13  Nm             6385 non-null   int32 
 14  rpm            6381 non-null   Int64 
dtypes: Int64(1), int32(2), int64(3), object(9)
memory usage: 704.7+ KB


In [77]:
data['mileage'] = data['mileage'].str.replace('km/kg','')
data['mileage'] = data['mileage'].astype(float)
data['engine'] = data['engine'].astype(int)
data['max_power'] = data['max_power'].astype(float)
data

name  year  selling_price  km_driven  \
0                Maruti Swift Dzire VDI  2014         450000     145500   
1          Skoda Rapid 1.5 TDI Ambition  2014         370000     120000   
2             Hyundai i20 Sportz Diesel  2010         225000     127000   
3                Maruti Swift VXI BSIII  2007         130000     120000   
4         Hyundai Xcent 1.2 VTVT E Plus  2017         440000      45000   
...                                 ...   ...            ...        ...   
6380  Maruti Wagon R VXI BS IV with ABS  2013         260000      50000   
6381         Hyundai i20 Magna 1.4 CRDi  2014         475000      80000   
6382                  Hyundai i20 Magna  2013         320000     110000   
6383              Hyundai Verna CRDi SX  2007         135000     119000   
6384             Maruti Swift Dzire ZDi  2009         382000     120000   

        fuel seller_type transmission                 owner  mileage  engine  \
0     Diesel  Individual       Manual           First Owner    23.40    1248   
1     Diesel  Individual       Manual          Second Owner    21.14    1498   
2     Diesel  Individual       Manual           First Owner    23.00    1396   
3     Petrol  Individual       Manual           First Owner    16.10    1298   
4     Petrol  Individual       Manual           First Owner    20.14    1197   
...      ...         ...          ...                   ...      ...     ...   
6380  Petrol  Individual       Manual          Second Owner    18.90     998   
6381  Diesel  Individual       Manual          Second Owner    22.54    1396   
6382  Petrol  Individual       Manual           First Owner    18.50    1197   
6383  Diesel  Individual       Manual  Fourth & Above Owner    16.80    1493   
6384  Diesel  Individual       Manual           First Owner    19.30    1248   

      max_power                     torque  seats   Nm   rpm  
0         74.00             190Nm@ 2000rpm      5  190  2000  
1        103.52        250Nm@ 1500-2500rpm      5  250  2000  
2         90.00   22.4 kgm at 1750-2750rpm      5  219  2250  
3         88.20      11.5@ 4,500(kgm@ rpm)      5  112  4500  
4         81.86          113.75nm@ 4000rpm      5  113  4000  
...         ...                        ...    ...  ...   ...  
6380      67.10              90Nm@ 3500rpm      5   90  3500  
6381      88.73      219.7Nm@ 1500-2750rpm      5  219  2125  
6382      82.85           113.7Nm@ 4000rpm      5  113  4000  
6383     110.00  24@ 1,900-2,750(kgm@ rpm)      5  235  2325  
6384      73.90             190Nm@ 2000rpm      5  190  2000  

[6385 rows x 15 columns]

In [78]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6385 entries, 0 to 6384
Data columns (total 15 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           6385 non-null   object 
 1   year           6385 non-null   int64  
 2   selling_price  6385 non-null   int64  
 3   km_driven      6385 non-null   int64  
 4   fuel           6385 non-null   object 
 5   seller_type    6385 non-null   object 
 6   transmission   6385 non-null   object 
 7   owner          6385 non-null   object 
 8   mileage        6385 non-null   float64
 9   engine         6385 non-null   int32  
 10  max_power      6385 non-null   float64
 11  torque         6385 non-null   object 
 12  seats          6385 non-null   int32  
 13  Nm             6385 non-null   int32  
 14  rpm            6381 non-null   Int64  
dtypes: Int64(1), float64(2), int32(3), int64(3), object(6)
memory usage: 679.8+ KB


In [79]:
power = (data.Nm*data.rpm)
data['hp'] = (power/7127)
data['hp']

0       53.318367
1       70.155746
2       69.138487
3       70.716992
4       63.420794
          ...    
6380     44.19812
6381     65.29746
6382    63.420794
6383    76.662691
6384    53.318367
Name: hp, Length: 6385, dtype: Float64

In [80]:
#iam still...

In [81]:
carlist = data['name']
carlist

0                  Maruti Swift Dzire VDI
1            Skoda Rapid 1.5 TDI Ambition
2               Hyundai i20 Sportz Diesel
3                  Maruti Swift VXI BSIII
4           Hyundai Xcent 1.2 VTVT E Plus
                      ...                
6380    Maruti Wagon R VXI BS IV with ABS
6381           Hyundai i20 Magna 1.4 CRDi
6382                    Hyundai i20 Magna
6383                Hyundai Verna CRDi SX
6384               Maruti Swift Dzire ZDi
Name: name, Length: 6385, dtype: object

In [82]:
#car = input("Enter car name =  ")
car = carlist[3]
result = process.extractOne(car,carlist)

result
i =result[2]
print(f'your selected car is       \n               =  {result[0]}  ')
print(f'hp             =  { data.hp[i]}  ')
print(f'rpm            =  { data.rpm[i]}  ')
print(f'nm             =  { data.Nm[i]}  ')
print(f'bhp             =  { data.max_power[i]}  ')

your selected car is       
               =  Maruti Swift VXI BSIII  
hp             =  70.716991721622  
rpm            =  4500  
nm             =  112  
bhp             =  88.2  


In [83]:
24 *9.80665

235.3596

In [84]:
# save data
data.to_csv("cleaned")

AttributeError: 'DataFrame' object has no attribute 't'

In [ ]:
df = data.drop(columns={'torque','rpm','hp'})
#df = data.drop(columns={'torque','rpm'}) 
df.info()

In [ ]:
df[df['mileage'] <= 10]
df[df['Nm'] <= 50]

In [ ]:
df= df.dropna()

In [ ]:
df.info()

In [ ]:
df["Nm"] = df['Nm'].astype(int)

# visualize Data

In [ ]:
import seaborn as sns
import matplotlib.pyplot as plt

In [ ]:
# check corrlation
corr =df.corr()
corr.style.background_gradient(cmap='coolwarm')

In [ ]:
sns.set_theme(style="darkgrid")
sns.countplot(df.owner)

In [ ]:
#Format the plot background and scatter plots for all the variables
sns.set_style(style='darkgrid')
sns.pairplot(df)

In [ ]:
sns.countplot(df['seller_type'])




In [ ]:
sns.countplot(df.transmission)

In [ ]:

sns.countplot(df['fuel'])

In [ ]:
sns.scatterplot(data=df,x = 'max_power',y='selling_price',hue='year')

### create a list of car names 

In [ ]:
import joblib
df1 = df
carlist = df1.name
carlist = carlist.unique()

filename = 'carlist.sav'
joblib.dump(carlist, filename)

In [ ]:
f_type = df1.fuel.unique()
sell_type = df1.seller_type.unique()
trans = df1.transmission.unique()
o_type = df1.owner.unique()

# DATA PREPROCESSING

In [ ]:
from sklearn.preprocessing import OrdinalEncoder,LabelEncoder
df.columns

In [ ]:
#CREATE OBJECTS FOR ALL CATEGORICAL DATA
le_name = LabelEncoder()
df['name']= le_name.fit_transform(df['name'])
df['name'].unique()

In [ ]:
le_fuel = LabelEncoder()
df['fuel']= le_fuel.fit_transform(df['fuel'])

le_trans = LabelEncoder()
df['transmission']= le_trans.fit_transform(df['transmission'])

le_owner = LabelEncoder()
df['owner']= le_owner.fit_transform(df['owner'])

le_sell = LabelEncoder()
df['seller_type']= le_sell.fit_transform(df['seller_type'])

In [ ]:
df['fuel'].unique()  ,df['transmission'].unique() ,df['owner'].unique() ,df['seller_type'].unique()

# outlier detection

In [ ]:
from scipy import stats
import numpy as np

# Calculate the z-scores
z_scores = stats.zscore(df)
z_scores

In [ ]:
# Convert to absolute values
abs_z_scores = np.abs(z_scores)

# Select data points with a z-scores above or below 3
filtered_entries = (abs_z_scores < 3).all(axis=1)

# Filter the dataset
df = df[filtered_entries]
df.shape

In [ ]:
df.info()

In [ ]:
df.describe()

### SPLITTING DATA

In [ ]:
y = df.selling_price

In [ ]:
x =df.drop(columns='selling_price')

In [ ]:
#df['hp'].isna().sum()

In [ ]:
#split data train test
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x,y,test_size=0.25)


In [ ]:
print(x_test.shape)
print(y_test.shape)
print(x_train.shape)
print(y_train.shape)

# creating multiple model

In [ ]:
from sklearn.linear_model import LinearRegression, Ridge,Lasso,ElasticNet
from sklearn.metrics import mean_squared_error

### LinearRegression

In [ ]:
lr = LinearRegression()
m_lr =lr.fit(x_train,y_train)
test_score = lr.score(x_test,y_test)
train_score = lr.score(x_train,y_train) 
print(train_score)
print(test_score)

In [ ]:
y_pred = m_lr.predict(x_test)
np.sqrt(mean_squared_error(y_pred,y_test))

### Ridge

In [ ]:
rg = Ridge()
m_rg =rg.fit(x_train,y_train)
test_score = rg.score(x_test,y_test)
train_score = rg.score(x_train,y_train) 
print(train_score)
print(test_score)

In [ ]:
y_pred = m_rg.predict(x_test)
np.sqrt(mean_squared_error(y_pred,y_test))

###  Lasso

In [ ]:
ls = Lasso()
m_ls =ls.fit(x_train,y_train)
test_score = ls.score(x_test,y_test)
train_score = ls.score(x_train,y_train) 
print(train_score)
print(test_score)

In [ ]:
y_pred = m_ls.predict(x_test)
np.sqrt(mean_squared_error(y_pred,y_test))

### ElasticNet

In [ ]:
el =ElasticNet()
m_el =el.fit(x_train,y_train)
test_score = el.score(x_test,y_test)
train_score = el.score(x_train,y_train) 
print(train_score)
print(test_score)


In [ ]:
y_pred = m_el.predict(x_test)
np.sqrt(mean_squared_error(y_pred,y_test))

### DecisionTreeRegressor

In [ ]:

from math import sqrt
from sklearn.tree import DecisionTreeRegressor
dt = DecisionTreeRegressor()
m_dt = dt.fit(x_train,y_train)
test_score =dt.score(x_test,y_test)
train_score = dt.score(x_train,y_train)
print(train_score)
print(test_score)
y_pred = m_dt.predict(x_test)
score = dt.score(x,y)
print(score)
sqrt(mean_squared_error(y_pred,y_test))

###  Hyper parameter tuning for Decision Tree

In [ ]:
from sklearn.model_selection import RandomizedSearchCV,GridSearchCV,KFold
# Define the Estimator and the Parameter Ranges
dt=DecisionTreeRegressor()
number_of_iterations = 5
min_samples_split = [2,5,7,10]
max_depth = range(2,30)
min_samples_leaf = [1,2,5]
# Define the param distribution dictionary
param_distributions = dict(min_samples_split=min_samples_split,
                           max_depth =max_depth,
                           min_samples_leaf =min_samples_leaf,
                          )

# Build the gridsearch
grid = RandomizedSearchCV(estimator=dt, 
                          param_distributions=param_distributions, 
                          n_iter=number_of_iterations, 
                          cv = 5)

grid_results = grid.fit(x_train, y_train)

# Summarize the results in a readable format
print("Best params: {0}, using {1}".format(grid_results.cv_results_['mean_test_score'], grid_results.best_params_))
results_df = pd.DataFrame(grid_results.cv_results_)

In [ ]:
result= grid_results.best_params_
min_samples_split = result.get("min_samples_split")
min_samples_leaf = result.get("min_samples_leaf")
max_depth = result.get('max_depth')

In [ ]:
dt = DecisionTreeRegressor(min_samples_split=min_samples_split,min_samples_leaf=min_samples_leaf,max_depth=max_depth)
dt_model = dt.fit(x_train,y_train)
test_score =dt.score(x_test,y_test)
train_score = dt.score(x_train,y_train)
print(train_score)
print(test_score)
score = dt.score(x,y)
print(score)
y_pred = dt_model.predict(x_test)
sqrt(mean_squared_error(y_pred,y_test))

## using assemble technique

In [ ]:

from sklearn.ensemble import BaggingRegressor,RandomForestRegressor
bag = BaggingRegressor()
rf  = RandomForestRegressor()


### BaggingRegressor

In [ ]:
from math import sqrt
from sklearn.metrics import mean_squared_error
m_bag = bag.fit(x_train,y_train)
test_score =bag.score(x_test,y_test)
train_score = bag.score(x_train,y_train)
print(train_score)
print(test_score)
y_pred = m_bag.predict(x_test)
sqrt(mean_squared_error(y_pred,y_test))

###  RandomForestRegressor

In [ ]:
m_rf = rf.fit(x_train,y_train)
test_score =rf.score(x_test,y_test)
train_score = rf.score(x_train,y_train)
print(train_score)
print(test_score)
y_pred = m_rf.predict(x_test)
print(sqrt(mean_squared_error(y_pred,y_test)))
y_pred = m_rf.predict(x)
sqrt(mean_squared_error(y_pred,y))

In [ ]:
m_rf.feature_importances_

In [ ]:
g=plt.scatter(y, y_pred)
g.axes.set_yscale('log')
g.axes.set_xscale('log')
g.axes.set_xlabel('True Values ')
g.axes.set_ylabel('Predictions ')
g.axes.axis('equal')
#.axes.axis('square')

### hyperparameter tuning for random forest

In [ ]:
# Define the Estimator and the Parameter Ranges
rf = RandomForestRegressor()
number_of_iterations = 10
min_samples_split = range(5,50)
max_depth = range(5,25)
max_features = range(5,13)
n_estimators = range(50,200)

# Define the param distribution dictionary
param_distributions = dict(min_samples_split=min_samples_split, 
                           max_depth=max_depth,
                           max_features=max_features,
                           n_estimators=n_estimators)

# Build the gridsearch
grid = RandomizedSearchCV(estimator=rf, 
                          param_distributions=param_distributions, 
                          n_iter=number_of_iterations, 
                          cv = 5)

grid_results = grid.fit(x_train, y_train)

# Summarize the results in a readable format
print("Best params: {0}, using {1}".format(grid_results.cv_results_['mean_test_score'], grid_results.best_params_))
results_df = pd.DataFrame(grid_results.cv_results_)


In [ ]:
result = grid_results.best_params_
n_estimators =result.get("n_estimators")
min_samples_split =result.get("min_samples_split")
max_features =result.get("max_features")
max_depth =result.get("max_depth")

### REtraining model using random forest

In [ ]:
x_train,x_test,y_train,y_test =train_test_split(x,y,train_size=0.80)


In [ ]:
rf = RandomForestRegressor(n_estimators=n_estimators,min_samples_split=min_samples_split,
                           max_features=max_features,max_depth=max_depth,random_state=42)
model = rf.fit(x_train,y_train)


In [ ]:
print(model.score(x_test,y_test))
print(model.score(x_train,y_train))
print(model.score(x,y))

In [ ]:
y_pred = model.predict(x_test)
sqrt(mean_squared_error(y_pred,y_test))

In [ ]:
y_pred = model.predict(x)
sqrt(mean_squared_error(y_pred,y))

In [ ]:
g=plt.scatter(y, y_pred)
g.axes.set_yscale('log')
g.axes.set_xscale('log')
g.axes.set_xlabel('True Values ')
g.axes.set_ylabel('Predictions ')
g.axes.axis('equal')
#.axes.axis('square')

In [ ]:
model.feature_importances_

In [ ]:
x.columns[5]

In [ ]:
input_data = np.array([['Mahindra KUV 100 mFALCON G80 K8 5str','2016','40000','Petrol','Individual','Manual','First Owner','18.15','1198','82','115','5']])
input_data

In [ ]:
input_data[:,0] = le_name.transform(input_data[:,0])
input_data[:,3] = le_fuel.transform(input_data[:,3])
input_data[:,4] = le_sell.transform(input_data[:,4])
input_data[:,5] = le_trans.transform(input_data[:,5])
input_data[:,6] = le_owner.transform(input_data[:,6])

In [ ]:
input_data = input_data.astype(float)

In [ ]:
y_pred = model.predict(input_data)

In [ ]:
import pickle
with open("cars.pkl","wb") as f:
    pickle.dump(carlist,f)

In [ ]:
with open("cars.pkl","rb") as f:
    cars =pickle.load(f)

In [ ]:
import pickle
data ={"model" :model,"le_name" :le_name,"le_fuel" :le_fuel,"le_sell" :le_sell,"le_trans":le_trans,"le_owner":le_owner}
with open("model.pkl","wb") as f:
    pickle.dump(data,f)
    

In [ ]:
def load_model():
    with open('model.pkl', 'rb') as file:
        data = pickle.load(file)
    return data
data = load_model()
model = data["model"]
le_name = data["le_name"]
le_fuel = data["le_fuel"]
le_sell = data["le_sell"]
le_trans= data["le_trans"]
le_owner= data["le_owner"]

In [ ]:
cars

In [ ]:
name = "Skoda Rapid 1.5 TDI Ambition"
result = process.extractOne(name,cars)
result[0]

In [ ]:
from fuzzywuzzy import process
name = input('inter car name here =  ')
result = process.extractOne(name,carlist)
name = result[2]

year = int(input("enter year =  "))

km_driven = int(input("enter km driven =  "))

fuel = input("enter fuel type =  ")
result = process.extractOne(fuel,f_type)
fuel = result[2]

seller_type = input("enter seller type =  ")
result = process.extractOne(seller_type,sell_type)
seller_type = result[2]

transmision = input("enter transmission type =  ")
result = process.extractOne(transmision,trans)
transmision = result[2]

owner = input("enter owner type =  ")
result = process.extractOne(owner,o_type)
owner = result[2]

mileage = float(input("enter mileage =  "))

engine = int(input("enter engine in cc  =  "))

max_power = float(input("enter maximum power =  "))

seats = int(input("enter seat size =  "))

Nm = int(input("enter torque genarated =  "))


In [ ]:
input_data = np.array([[name,year,km_driven,fuel,seller_type,transmision,owner,mileage,engine,max_power,seats,Nm]])
input_data[:,0] = le_name.transform(input_data[:,0])
input_data[:,3] = le_fuel.transform(input_data[:,3])
input_data[:,4] = le_sell.transform(input_data[:,4])
input_data[:,5] = le_trans.transform(input_data[:,5])
input_data[:,6] = le_owner.transform(input_data[:,6])
input_data = input_data.astype(float)
y_pred = model.predict(input_data)
print(y_pred)

In [ ]:
x.columns

In [ ]:
f_type

In [ ]:
sell_type

In [ ]:
o_type

In [ ]:
trans

In [ ]:
x.seats.unique()

In [ ]:
carlist

# feature scalling

from sklearn.preprocessing import StandardScaler
scaller = StandardScaler()

X  = scaller.fit_transform(x)

x_train,x_test,y_train,y_test =train_test_split(X,y,train_size=0.80)
rf = RandomForestRegressor(n_estimators=n_estimators,min_samples_split=min_samples_split,
                           max_features=max_features,max_depth=max_depth,random_state=42)
model = rf.fit(x_train,y_train)
print(model.score(x_test,y_test))
print(model.score(x_train,y_train))
print(model.score(X,y))

y_pred = model.predict(x_test)
sqrt(mean_squared_error(y_pred,y_test))

y_pred = model.predict(X)
sqrt(mean_squared_error(y_pred,y))


model.feature_importances_

In [ ]:
x

In [ ]:
print(le_name.inverse_transform([1188]))


In [ ]:
print(le_fuel.inverse_transform([1]))
print(le_sell.inverse_transform([1]))
print(le_trans.inverse_transform([1]))
print(le_owner.inverse_transform([0]))